src/main/scala/mesosphere/marathon/storage/migration/MigrationTo160.scala

```diff
@@ -0,0 +1,127 @@
+    */
+  def migrateReservations(instanceRepository: InstanceRepository, persistenceStore: PersistenceStore[_, _, _])(implicit ctx: ExecutionContext, mat: Materializer): Future[Done] = {
+
+    logger.info("Starting reservations migration to 1.6.0")
+
+    /**
+      * We're trying to find if we have a ZooKeeper store because it provides objects as byte arrays and this
+      * makes serialization into json easier.
+      */
+    val maybeStore: Option[ZkPersistenceStore] = {
+
+      def findZkStore(ps: PersistenceStore[_, _, _]): Option[ZkPersistenceStore] = {
+        ps match {
+          case zk: ZkPersistenceStore =>
+            Some(zk)
+          case lcps: LazyCachingPersistenceStore[_, _, _] =>
+            findZkStore(lcps.store)
+          case lvcps: LazyVersionCachingPersistentStore[_, _, _] =>
+            findZkStore(lvcps.store)
+          case ltcps: LoadTimeCachingPersistenceStore[_, _, _] =>
+            findZkStore(ltcps.store)
+          case other =>
+            None
+        }
+      }
+
+      findZkStore(persistenceStore)
+    }
+
+    implicit val instanceResolver: IdResolver[Instance.Id, JsValue, String, ZkId] =
+      new IdResolver[Instance.Id, JsValue, String, ZkId] {
+        override def toStorageId(id: Id, version: Option[OffsetDateTime]): ZkId =
+          ZkId(category, id.idString, version)
+
+        override val category: String = "instance"
+
+        override def fromStorageId(key: ZkId): Id = Instance.Id(key.id)
+
+        override val hasVersions: Boolean = false
+
+        override def version(v: JsValue): OffsetDateTime = OffsetDateTime.MIN
+      }
+
+    implicit val instanceJsonUnmarshaller: Unmarshaller[ZkSerialized, JsValue] =
+      Unmarshaller.strict {
+        case ZkSerialized(byteString) =>
+          Json.parse(byteString.utf8String)
+      }
+
+    import Reservation.reservationFormat
+    import Instance.instanceJsonReads
+
+    maybeStore.map { store =>
+      instanceRepository
+        .ids()
+        .mapAsync(1) { instanceId =>
+          store.get(instanceId)
+        }
+        .mapConcat {
+          case Some(jsValue) =>
+            val instance = jsValue.as[Instance]
+            // Prior to Marathon 1.6.0, persistent volumes are supported only with apps,
+            // therefore reservation objects can only appear in app instances, and since
+            // an app has only one task by definition, there is only one KV pair in a taskMap
+            // object.
+            //
+            // We use .headOption here to handle the case of apps with no persistent volumes.
+            val maybeReservationJson = (jsValue \ "tasksMap" \\ "reservation").headOption
+
+            maybeReservationJson.map { reservationJson =>
+              reservationJson.as[Reservation] -> instance :: Nil
+            } getOrElse {
+              Nil
+            }
+
+          case _ => Nil
+        }
+        .mapConcat { case (reservation, instance) =>
```

Comment:
I thought we get a list from the `mapConcat` before. Why do we have a reservation and an instance here?

Reply:
mapConcat unwraps list and emits its elements downstream one by one

Reply:
What do you think about making these to lambdas helper methods? This makes reading and testing much simpler.